src/app/ReadClient.cpp

```diff
@@ -0,0 +1,298 @@
+                                     Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse));
+    SuccessOrExit(err);
+    MoveToState(kState_ReadRequestSending);
+
+exit:
+    ChipLogFunctError(err);
+    return err;
+}
+
+void ReadClient::OnMessageReceived(Messaging::ExchangeContext * apExchangeContext, const PacketHeader & aPacketHeader,
+                                   const PayloadHeader & aPayloadHeader, System::PacketBufferHandle aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    VerifyOrDie(apExchangeContext == mpExchangeCtx);
+
+    ClearExistingExchangeContext();
+
+    err = ProcessReportData(std::move(aPayload));
+    SuccessOrExit(err);
+
+exit:
+    Reset();
+    return;
+}
+
+CHIP_ERROR ReadClient::ClearExistingExchangeContext()
+{
+    if (mpExchangeCtx != nullptr)
+    {
+        mpExchangeCtx->Abort();
+        mpExchangeCtx = nullptr;
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ReadClient::ProcessReportData(System::PacketBufferHandle && aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::app::InteractionModelDelegate::InEventParam inParam;
+    chip::app::InteractionModelDelegate::OutEventParam outParam;
+    ReportData::Parser report;
+
+    bool isEventListPresent = false;
+    bool suppressResponse   = false;
+
+    chip::System::PacketBufferTLVReader reader;
+
+    if (nullptr != mpDelegate)
+    {
+        inParam.mReportDataRequest.exchangeContext = mpExchangeCtx;
+        mpDelegate->HandleEvent(chip::app::InteractionModelDelegate::EventId::kReportDataRequest, inParam, outParam);
+    }
+
+    reader.Init(std::move(aPayload));
+    reader.Next();
+
+    err = report.Init(reader);
+    SuccessOrExit(err);
+
+    err = report.CheckSchemaValidity();
+    SuccessOrExit(err);
+
+    err = report.GetSuppressResponse(&suppressResponse);
+    if (CHIP_END_OF_TLV == err)
+        err = CHIP_NO_ERROR;
+    SuccessOrExit(err);
+
+    err = report.GetMoreChunkedMessages(&mMoreChunkedMessages);
+    if (CHIP_END_OF_TLV == err)
+        err = CHIP_NO_ERROR;
+    SuccessOrExit(err);
+
+    {
+        EventList::Parser eventList;
+
+        err = report.GetEventDataList(&eventList);
+        if (CHIP_NO_ERROR == err)
+        {
+            isEventListPresent = true;
+        }
+        else if (CHIP_END_OF_TLV == err)
+        {
+            isEventListPresent = false;
+            err                = CHIP_NO_ERROR;
+        }
+        SuccessOrExit(err);
+
+        if (isEventListPresent && nullptr != mpDelegate)
+        {
+            // re-initialize the reader (reuse to save stack depth).
+            inParam.Clear();
+            outParam.Clear();
+            eventList.GetReader(&reader);
+            inParam.mEventStreamReceived.reader = &reader;
+            mpDelegate->HandleEvent(chip::app::InteractionModelDelegate::EventId::kEventStreamReceived, inParam, outParam);
+        }
+    }
+
+    if (!suppressResponse)
+    {
+        /*
```

Comment:
It would be relatively large PR for this status report support in IM, prefer add it in next PR.

Just add TODO with more info, Add status report support and correspond handler in ReadHandler, particular for situation send multiple status reporrts when there are multiple reports